theano/tensor/opt.py

```diff
@@ -2591,6 +2591,64 @@ def local_adv_sub1_adv_inc_sub1(node):
     return [T.cast(y, node.outputs[0].dtype)]
 
 
+@register_specialize
+@register_stabilize
+@register_canonicalize
+@gof.local_optimizer([IncSubtensor,
+                      AdvancedIncSubtensor,
+                      AdvancedIncSubtensor1])
+def local_useless_incsubtensor_alloc(node):
+    """
+    Replaces an [Advanced]IncSubtensor[1], whose increment is an `alloc` of
+    a variable that has either broadcastable or matching dimensions with the
+    target variable that shall be incremented, by one that skips the
+    intermediate `alloc`.
+    """
+    if isinstance(node.op, (IncSubtensor,
+                            AdvancedIncSubtensor,
+                            AdvancedIncSubtensor1)):
+        x = node.inputs[0]
+        y = node.inputs[1]
+        i = node.inputs[2:]
+
+        if y.owner is not None and isinstance(y.owner.op, T.Alloc):
+            # z is the input of the Alloc op, i.e. T.alloc(z, <shape>)
+            z = y.owner.inputs[0]
+
+            # A non-empty list of conditions means that `z` has some
+            # non-broadcastable dimensions that cannot be checked statically
+            # for equal size with the corresponding dimension of `x`. Thus, a
+            # runtime check is necessary.
+            try:
+                cond = [T.eq(x.shape[k], z.shape[k])
```

Comment:
You are right, the condition is more complicated. There is the case you mention. I think that we need to make explicit the implicit padding of y when y.ndim != the output of the alloc.ndim. I taught of a way to work around the problem you told with the same_shape. Compare again y directly. Here is new pseudo-code:

```
z = y.owner.inputs[0]
# I explicitly build the z_shape and z_broad to include the extra implicit dimensions. This ease the algo bellow.
z_shape = (1, ) * (y.ndim - z.ndim) + z.shape
z_broad = (True, ) * (y.ndim - z.ndim) + z.broadcastable
assert x.ndim = len(z_shape)
cond = [T.eq(x.shape[k], z_shape[k])
    for k in xrange(y.ndim)
    if not z_broad[k] and
# I compare x and y to not mess with how the alloc output shape is computed.
# I let the Alloc.infer_shape tell how this is done.
    not node.fgraph.shape_feature.same_shape(x, y,
    dim_x=k,
    dim_y=k)]
```